packages/core/plugins/core-plugins-server-internal/src/plugins_system.ts

```diff
@@ -298,67 +298,74 @@ export class PluginsSystem<T extends PluginType> {
     return publicPlugins;
   }
 
-  /**
-   * Gets topologically sorted plugin names that are registered with the plugin system.
-   * Ordering is possible if and only if the plugins graph has no directed cycles,
-   * that is, if it is a directed acyclic graph (DAG). If plugins cannot be ordered
-   * an error is thrown.
-   *
-   * Uses Kahn's Algorithm to sort the graph.
-   */
   private getTopologicallySortedPluginNames() {
-    // We clone plugins so we can remove handled nodes while we perform the
-    // topological ordering. If the cloned graph is _not_ empty at the end, we
-    // know we were not able to topologically order the graph. We exclude optional
-    // dependencies that are not present in the plugins graph.
-    const pluginsDependenciesGraph = new Map(
-      [...this.plugins.entries()].map(([pluginName, plugin]) => {
-        return [
-          pluginName,
-          new Set([
-            ...plugin.requiredPlugins,
-            ...plugin.optionalPlugins.filter((dependency) => this.plugins.has(dependency)),
-          ]),
-        ] as [PluginName, Set<PluginName>];
-      })
-    );
-
-    // First, find a list of "start nodes" which have no outgoing edges. At least
-    // one such node must exist in a non-empty acyclic graph.
-    const pluginsWithAllDependenciesSorted = [...pluginsDependenciesGraph.keys()].filter(
-      (pluginName) => pluginsDependenciesGraph.get(pluginName)!.size === 0
-    );
-
-    const sortedPluginNames = new Set<PluginName>();
-    while (pluginsWithAllDependenciesSorted.length > 0) {
-      const sortedPluginName = pluginsWithAllDependenciesSorted.pop()!;
-
-      // We know this plugin has all its dependencies sorted, so we can remove it
-      // and include into the final result.
-      pluginsDependenciesGraph.delete(sortedPluginName);
-      sortedPluginNames.add(sortedPluginName);
-
-      // Go through the rest of the plugins and remove `sortedPluginName` from their
-      // unsorted dependencies.
-      for (const [pluginName, dependencies] of pluginsDependenciesGraph) {
-        // If we managed delete `sortedPluginName` from dependencies let's check
-        // whether it was the last one and we can mark plugin as sorted.
-        if (dependencies.delete(sortedPluginName) && dependencies.size === 0) {
-          pluginsWithAllDependenciesSorted.push(pluginName);
-        }
-      }
+    if (!this.sortedPluginNames) {
+      this.sortedPluginNames = getTopologicallySortedPluginNames(this.plugins);
```

Comment:
NIT Could we clean this cache whenever we `addPlugin()`? I know it shouldn't happen, but just for the peace of mind.